src/main/java/seedu/address/logic/parser/FindTransactionCommandParser.java

```diff
@@ -0,0 +1,107 @@
+
+    public static final String BUY_KEYWORD = "buy";
+    public static final String SELL_KEYWORD = "sell";
+
+    /**
+     * defines the type of transaction
+     */
+    public enum TransactionType {
+        BUY_TRANSACTION, SELL_TRANSACTION, EMPTY
+    }
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the FindTransactionCommand
+     * and returns a FindTransactionCommand object for execution.
+     *
+     * @throws ParseException if the user input does not conform the expected format
+     */
+    public FindTransactionCommand parse(String args) throws ParseException {
+        requireNonNull(args);
+
+        // find transaction by type of transaction, name of supplier, name of good.
+        // or any combination of the above
+        // more factor -> more constrain
+        // if there are multiple input for name, good name, all inputs will be taken.
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NAME,
+                PREFIX_GOOD_NAME);
+
+        // preamble stores the type of transaction
+        TransactionType typeOfTransaction = parseTypeOfTransaction(argMultimap);
+
+        // name stores the name of supplier
+        String[] supplierNameKeywords = setKeywords(argMultimap, PREFIX_NAME);
+
+        // good name stores the name of the good
+        String[] goodNameKeywords = setKeywords(argMultimap, PREFIX_GOOD_NAME);
+
+        // at least one field must be provided
+        if (typeOfTransaction.equals(TransactionType.EMPTY)
+                && supplierNameKeywords.length == 0 && goodNameKeywords.length == 0) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
+                    FindTransactionCommand.MESSAGE_NO_FIELD_PROVIDED));
+        }
+
+        return new FindTransactionCommand(
+                new TransactionContainKeywordsPredicate(typeOfTransaction,
+                        Arrays.asList(supplierNameKeywords), Arrays.asList(goodNameKeywords)));
+    }
+
+    /**
+     * parses values in the prefix
+     * @param argMultimap stores the input
+     * @param prefix can be supplier name or good name
+     * @return arrays of the individual words in the input for particular prefix, if the input is empty (which means
+     * this prefix is unspecified, return an empty array
+     * @throws ParseException if the prefix is given, but the value is empty
+     */
+    private String[] setKeywords(ArgumentMultimap argMultimap, Prefix prefix) throws ParseException {
+        Optional<String> words = argMultimap.getValue(prefix);
+        if (words.isPresent()) {
+            if (words.get().equals("")) {
+                throw new ParseException(
+                        String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindTransactionCommand.EMPTY_VALUE_WITH_PREFIX));
+            }
+            return words.get().split("\\s+");
+        }
+        return new String[0];
```

Comment:
Do you think this is a good candidate for "removing the bad cases" style of code? E.g.:
```
if (!words.isPresent()) {
    return new String[0];
}


// words will be present if it makes it to this line already
if (words.get().equals("")) {
    // throw parse exception
}

return words.get().split("\\s+");
```
That removes the nested if statements too!

Reply:
thats a good strategy, thanks.